src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWalker.java

```diff
@@ -0,0 +1,305 @@
+        }
+    }
+
+    public long memoryUsage()
+    {
+        return state.memoryUsage;
+    }
+
+    @Override
+    public void close()
+    {
+        FileUtils.closeQuietly(treeIndexFile);
+    }
+
+    void traverse(TraversalCallback callback)
+    {
+        state.reset();
+        traverse(callback, new IntArrayList());
+    }
+
+    private void traverse(TraversalCallback callback, IntArrayList pathToRoot)
+    {
+        if (state.isLeafNode())
+        {
+            // In the unbalanced case it's possible the left most node only has one child:
+            if (state.nodeExists())
+            {
+                callback.onLeaf(state.nodeID, state.getLeafBlockFP(), pathToRoot);
+            }
+        }
+        else
+        {
+            IntArrayList currentPath = new IntArrayList();
+            currentPath.addAll(pathToRoot);
+            currentPath.add(state.nodeID);
+
+            state.pushLeft();
+            traverse(callback, currentPath);
+            state.pop();
+
+            state.pushRight();
+            traverse(callback, currentPath);
+            state.pop();
+        }
+    }
+
+    interface TraversalCallback
+    {
+        void onLeaf(int leafNodeID, long leafBlockFP, IntArrayList pathToRoot);
+    }
+
+    /**
+     * This maintains the state for a traversal of the packed index. It is loaded once and can be resused
+     * by calling the reset method.
+     * <p>
+     * The packed index is a packed representation of a balanced tree and takes the form of a packed array of
+     * file pointer / split value pairs. Both the file pointers and split values are prefix compressed by tree level
+     * requiring us to maintain a stack of values for each level in the tree. The stack size is always the tree depth.
+     * <p>
+     * The tree is traversed by recursively following the left and then right subtrees under the current node. For the
+     * following tree (split values in square brackets):
+     * <pre>
+     *        1[16]
+     *       / \
+     *      /   \
+     *     2[8]  3[24]
+     *    / \   / \
+     *   4   5 6   7
+     * </pre>
+     * The traversal will be 1 -> 2 -> 4 -> 5 -> 3 -> 6 -> 7 with nodes 4, 5, 6 & 7 being leaf nodes.
+     * <p>
+     * Assuming the full range of values in the tree is 0 -> 32, the non-leaf nodes will represent the following
+     * values:
+     * <pre>
+     *         1[0-32]
+     *        /      \
+     *    2[0-16]   3[16-32]
+     * </pre>
+     */
+    final static class TraversalState
+    {
+        final int bytesPerValue;
+        final int numLeaves;
+        final int treeDepth;
+        final byte[] minPackedValue;
+        final byte[] maxPackedValue;
+        final long pointCount;
+        final int maxPointsInLeafNode;
+        final long memoryUsage;
+
+        // used to read the packed index byte[]
+        final ByteArrayDataInput dataInput;
+        // holds the minimum (left most) leaf block file pointer for each level we've recursed to:
```

Comment:
When we say "for each level", what we mean is "for the sub-trees rooted at the nodes on our current path"?

Reply:
Correct. The stacks are slightly confusing and I have been trying to come up with a better explanation for them. They are very rooted to the traversal pattern.